### PR TITLE
Add TIMESTAMP and BOOLEAN data type support

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
## Description
Add `TIMESTAMP` and `BOOLEAN` data type support, and the primitives to support new data types stored as the supported types.
Internally `TIMESTAMP` is stored as `LONG` (millis since epoch), `BOOLEAN` is stored as `INT` (`0` for `false`, `1` for `true`).
In the query response, `TIMESTAMP` values follow the JDBC format (e.g. `2021-01-01 01:01:01.123`).

The source data of a `TIMESTAMP` field should be either long values (millis since epoch) or String values of JDBC timestamp format; The source data of a `BOOLEAN` field should be boolean, numbers (non-zero for `true`, zero for `false`), or string (`"true"` for `true`, `false` otherwise).

Support `TIMESTAMP` and `BOOLEAN` in:
- Data Ingestion
- Filtering
- Selection
- Distinct
- Order-by
- Aggregation
- Group-by
- Transform
- Having
- Post-Aggregation
- Scalar function (`java.sql.Timestamp` for `TIMESTAMP` and primitive `boolean` for `BOOLEAN`)

## Release Notes
`TIMESTAMP` and `BOOLEAN` is supported as the dimension/time field data type within the schema.
**`BOOLEAN` is no longer converted to `STRING` in the schema.**

## Documentation
To be added to https://docs.pinot.apache.org/basics/components/schema#data-types

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java [hunk 1]
```diff
@@ -215,7 +236,34 @@ public int hashCode() {
   }
 
   public enum ColumnDataType {
-    INT, LONG, FLOAT, DOUBLE, STRING, BYTES, OBJECT, INT_ARRAY, LONG_ARRAY, FLOAT_ARRAY, DOUBLE_ARRAY, STRING_ARRAY;
+    INT,
+    LONG,
+    FLOAT,
+    DOUBLE,
+    BOOLEAN /* Stored as INT */,
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:243]: Should `BOOLEAN` should be stored as `BYTE` to save space?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java [hunk 2]
```diff
@@ -92,6 +115,16 @@ public double toDouble(Object value) {
       return ((Byte) value).doubleValue();
     }
 
+    @Override
+    public boolean toBoolean(Object value) {
+      return (Byte) value > 0;
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:120]: Can we replace this by `return (Byte) value != 0`? I think the common convention (at least in c/c++ and a few other languages) is to treat zero as false and everything else as true. Same with other data types (character, integer, long, float) toBoolean function.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java [hunk 3]
```diff
@@ -315,6 +408,58 @@ public Double convert(Object value, PinotDataType sourceType) {
...
+    }
+
+    @Override
+    public double toDouble(Object value) {
+      return ((Timestamp) value).getTime();
+    }
+
+    @Override
+    public boolean toBoolean(Object value) {
+      throw new UnsupportedOperationException("Cannot convert value from TIMESTAMP to BOOLEAN");
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:434]: I would have done zero as `false` and all other timestamp values as `true`. This convention generalizes to work with everything and type conversion becomes simple :-)

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java [hunk 4]
```diff
@@ -375,6 +530,16 @@ public double toDouble(Object value) {
       throw new UnsupportedOperationException("Cannot convert value from BYTES to DOUBLE");
     }
 
+    @Override
+    public boolean toBoolean(Object value) {
+      throw new UnsupportedOperationException("Cannot convert value from BYTES to BOOLEAN");
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:535]: If BYTES is not 0, it is true; otherwise, false?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java [hunk 5]
```diff
@@ -412,6 +577,16 @@ public double toDouble(Object value) {
       return ((Number) value).doubleValue();
     }
 
+    @Override
+    public boolean toBoolean(Object value) {
+      return ((Number) value).intValue() > 0;
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:582]: Can object be null? Would null Object be considered true or false? What if we add NULL value support later?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java [hunk 6]
```diff
@@ -775,6 +969,18 @@ public static PinotDataType getPinotDataTypeForIngestion(FieldSpec fieldSpec) {
         return fieldSpec.isSingleValueField() ? PinotDataType.FLOAT : PinotDataType.FLOAT_ARRAY;
       case DOUBLE:
         return fieldSpec.isSingleValueField() ? PinotDataType.DOUBLE : PinotDataType.DOUBLE_ARRAY;
+      case BOOLEAN:
+        if (fieldSpec.isSingleValueField()) {
+          return PinotDataType.BOOLEAN;
+        } else {
+          throw new IllegalStateException("There is no multi-value type for BOOLEAN");
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:976]: Any specific reason why no multi-value support for BOOLEAN? Boolean is a primitive type just like Numerical types?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java [hunk 7]
```diff
@@ -775,6 +969,18 @@ public static PinotDataType getPinotDataTypeForIngestion(FieldSpec fieldSpec) {
...
+        }
+      case TIMESTAMP:
+        if (fieldSpec.isSingleValueField()) {
+          return PinotDataType.TIMESTAMP;
+        } else {
+          throw new IllegalStateException("There is no multi-value type for TIMESTAMP");
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:982]: Same with Timestamp, would like to see multi-valued support here as well unless there is a very specific reason not to do so?

File: pinot-common/src/test/java/org/apache/pinot/common/data/DateTimeFormatSpecTest.java [hunk 8]
```diff
@@ -82,6 +86,8 @@ public void testFromMillisToFormat(String format, long timeMs, String expectedFo
     entries.add(new Object[]{"1:MILLISECONDS:EPOCH", 1498892400000L, "1498892400000"});
     entries.add(new Object[]{"1:HOURS:EPOCH", 0L, "0"});
     entries.add(new Object[]{"5:MINUTES:EPOCH", 1498892400000L, "4996308"});
+    entries.add(new Object[]{"1:MILLISECONDS:TIMESTAMP", Timestamp
```
Comment (amrishlal) [inline pinot-common/src/test/java/org/apache/pinot/common/data/DateTimeFormatSpecTest.java:89]: Would be good to see a test case for `1:NANOSECONDS:TIMESTAMP` specially since Pinot supports `NANOSECONDS` and `java.sql.Timestamp` seems to require a separate call to `java.sql.Timestamp.setNanos(...)` to set nanoseconds in timestamp.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/EqualsPredicateEvaluatorFactory.java [hunk 9]
```diff
@@ -54,30 +57,36 @@ public static BaseDictionaryBasedPredicateEvaluator newDictionaryBasedEvaluator(
    */
   public static BaseRawValueBasedPredicateEvaluator newRawValueBasedEvaluator(EqPredicate eqPredicate,
       DataType dataType) {
+    String value = eqPredicate.getValue();
     switch (dataType) {
       case INT:
-        return new IntRawValueBasedEqPredicateEvaluator(eqPredicate);
+        return new IntRawValueBasedEqPredicateEvaluator(Integer.parseInt(value));
```
Comment (amrishlal) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/EqualsPredicateEvaluatorFactory.java:63]: Are these changes related to BOOLEAN and TIMESTAMP support?

**Comment by amrishlal** [comment]
This is certainly good work. Both BOOLEAN and TIMESTAMP would be very useful. I haven't looked at all the files yet, but will look again later. Not sure if I missed them, but would like to see more SQL level test cases for these two data types.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:243]: Internally we don't have `BYTE` type, so here we use the smallest type to store `BOOLEAN`.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:120]: Good suggestion, changed

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:434]: Done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:535]: I wouldn't worry too much about that. People are not supposed to use `BYTES` as `BOOLEAN`, and throwing exception might be a safer option to avoid unexpected behavior.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java (hunk 5 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:582]: We won't get `null` here. (Usually we annotation arguments that can be `null` with `Nullable`)

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/EqualsPredicateEvaluatorFactory.java (hunk 9 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/EqualsPredicateEvaluatorFactory.java:63]: Yes, because `BOOLEAN` and `TIMESTAMP` is stored as another type, making these changes so that the code can be reused.

File: pinot-common/src/test/java/org/apache/pinot/common/data/DateTimeFormatSpecTest.java (hunk 8 above)
Comment (Jackie-Jiang) [reply pinot-common/src/test/java/org/apache/pinot/common/data/DateTimeFormatSpecTest.java:89]: SQL Timestamp always have values in `yyyy-MM-dd HH:mm:ss.SSS` format, and it doesn't allow `nanoseconds` granularity. The first 2 values are actually ignored (similar to date time format which only reads the 4th argument as the format). We should consider changing the order for these arguments for clarity, but that is out of the scope of this PR.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java (hunk 6 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:976]: I feel it should be very rare to have MV boolean field (there are only 2 possible values, and the dimension should be either `true` or `false`). If required, we can add the support later. Added a todo here.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java (hunk 7 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:982]: `TIMESTAMP` is mostly used as the time field, which does not allow MV. If required, we can add the support later. Added a todo here.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/TimestampUtils.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+   * </ul>
+   */
+  public static long toMillisSinceEpoch(String timestampString) {
+    try {
+      return Timestamp.valueOf(timestampString).getTime();
+    } catch (Exception e) {
+      try {
+        return Long.parseLong(timestampString);
+      } catch (Exception e1) {
+        throw new IllegalArgumentException(String.format("Invalid timestamp: '%s'", timestampString));
```
Comment (amrishlal) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/TimestampUtils.java:61]: Same here. Would like to see root exception being included.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/TimestampUtils.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+   * </ul>
+   */
+  public static Timestamp toTimestamp(String timestampString) {
+    try {
+      return Timestamp.valueOf(timestampString);
+    } catch (Exception e) {
+      try {
+        return new Timestamp(Long.parseLong(timestampString));
+      } catch (Exception e1) {
+        throw new IllegalArgumentException(String.format("Invalid timestamp: '%s'", timestampString));
```
Comment (amrishlal) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/TimestampUtils.java:41]: Can we also include the root exception in IllegalArgumentException. Makes it easier to trace back to the root cause or exception sequence.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 12]
```diff
@@ -737,11 +732,11 @@ public static void removeColumnMetadataInfo(PropertiesConfiguration properties,
...
    */
-  public static ForwardIndexCreator getRawIndexCreatorForColumn(File file,
-      ChunkCompressionType compressionType, String column, DataType dataType, int totalDocs,
-      int lengthOfLongestEntry, boolean deriveNumDocsPerChunk, int writerVersion)
+  public static ForwardIndexCreator getRawIndexCreatorForColumn(File file, ChunkCompressionType compressionType,
+      String column, DataType dataType, int totalDocs, int lengthOfLongestEntry, boolean deriveNumDocsPerChunk,
+      int writerVersion)
       throws IOException {
-    switch (dataType) {
+    switch (dataType.getStoredType()) {
```
Comment (amrishlal) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:739]: Can you please add some comments in DataType.java explaining the usage DataType vs stored DataType, i.e when should DataType be used and when should DataType.getStoredType() be used?

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/TimestampUtils.java (hunk 11 above)
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/TimestampUtils.java:41]: The only exception we can get here is `NumberFormatException` from `Long.parseLong()`, so IMO logging the invalid timestamp should be enough for debugging purpose.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java (hunk 12 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:739]: Added some comments in the javadoc of `DataType.getStoredType()`

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java (hunk 1 above)
Comment (yupeng9) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:243]: Curious, do we plan other int types in the future? e.g. tinyint, short, int etc.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java [hunk 13]
```diff
@@ -60,6 +63,16 @@ public double toDouble(Object value) {
       return ((Boolean) value) ? 1d : 0d;
     }
 
+    @Override
+    public boolean toBoolean(Object value) {
```
Comment (yupeng9) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:67]: shall this return type be `Boolean` to allow null value in future?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java [hunk 14]
```diff
@@ -315,6 +408,58 @@ public Double convert(Object value, PinotDataType sourceType) {
...
+    }
+
+    @Override
+    public double toDouble(Object value) {
+      return ((Timestamp) value).getTime();
+    }
+
+    @Override
+    public boolean toBoolean(Object value) {
+      return ((Timestamp) value).getTime() != 0;
```
Comment (yupeng9) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:434]: whats the meaning of this?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:243]: Probably no. It doesn't make sense to add them as logical types and backed by int because that won't save any storage.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java (hunk 13 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:67]: We won't allow `null` value inside this type conversion util class. `null` value should be handled outside.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java (hunk 14 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:434]: This method will be triggered when the input value is of type `Timestamp`, but the schema defines the field as `BOOLEAN`. In regular cases it should not be used.

**Comment by Jackie-Jiang** [comment]
> This is certainly good work. Both BOOLEAN and TIMESTAMP would be very useful. I haven't looked at all the files yet, but will look again later. Not sure if I missed them, but would like to see more SQL level test cases for these two data types.

@amrishlal Added `BooleanQueriesTest` and `TimestampQueriesTest` to test the query flow for various types of queries

**Comment by amrishlal** [comment]
> @amrishlal Added `BooleanQueriesTest` and `TimestampQueriesTest` to test the query flow for various types of queries

I would have probably used 20 records instead of 1000, but looks good to me :-)

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java [hunk 15]
```diff
@@ -352,13 +368,41 @@ public int hashCode() {
...
      * Returns the data type stored in Pinot.
+     * <p>Pinot internally stores data (physical) in INT, LONG, FLOAT, DOUBLE, STRING, BYTES type, other data types
+     * (logical) will be stored as one of these types.
+     * <p>Stored type should be used when reading the physical stored values from Dictionary, Forward Index etc.
      */
     public DataType getStoredType() {
-      return this == BOOLEAN ? STRING : this;
+      switch (this) {
+        case BOOLEAN:
+          return INT;
```
Comment (siddharthteotia) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:392]: Not necessarily in this PR, but would it make sense to store this is a bit / byte as opposed to 32 bits in future ? If there is a plan to do so, may be we can write a property to segment metadata (BOOLEAN_STORED_TYPE= INT) to support evolution

File: pinot-core/src/test/java/org/apache/pinot/queries/BooleanQueriesTest.java [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+      // Insert data in 6 different formats
+      switch (i % 6) {
+        case 0:
+          record.putValue(BOOLEAN_COLUMN, false);
+          break;
+        case 1:
+          record.putValue(BOOLEAN_COLUMN, 1);
+          break;
+        case 2:
+          record.putValue(BOOLEAN_COLUMN, 0L);
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/queries/BooleanQueriesTest.java:102]: (nit) might be helpful to add few notes on semantics of BOOLEAN in FieldSpec class just like you have handled multiple different formats here.

Is it correct to say that any signed numerical value > 0 is interpreted as true and everything else as false ?

Also, for source string values, will record reader (or some code in segment generator) throw exception if the source string value for a BOOLEAN data type column is anything apart from "true" or "false" ?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 17]
```diff
@@ -235,33 +234,32 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
...
       }
 
       if (textIndexColumns.contains(columnName)) {
         // Initialize text index creator
         Preconditions.checkState(fieldSpec.isSingleValueField(),
             "Text index is currently only supported on single-value columns");
-        Preconditions.checkState(fieldSpec.getDataType() == STRING,
-            "Text index is currently only supported on STRING type columns");
+        Preconditions
+            .checkState(storedType == DataType.STRING, "Text index is currently only supported on STRING type columns");
```
Comment (siddharthteotia) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:253]: With the current ongoing discussion for nested/json types, flattening etc, the short term will have JSON data type and the forward index will continue to have it as STRING as stored type. 
The above check will actually pass for a json column if the user wants to create a text index on it. I don't think we want that at least until we have tested how text index will work on json (stringified ) data. I think the check should still happen using the logical type and not the stored type
cc @amrishlal

File: pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+
+
+/**
+ * Queries test for TIMESTAMP data type.
+ */
+public class TimestampQueriesTest extends BaseQueriesTest {
+  private static final File INDEX_DIR = new File(FileUtils.getTempDirectory(), "TimestampQueriesTest");
+  private static final String RAW_TABLE_NAME = "testTable";
+  private static final String SEGMENT_NAME = "testSegment";
+  private static final long BASE_TIMESTAMP = Timestamp.valueOf("2021-01-01 00:00:00").getTime();
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java:59]: Adding a few notes on semantics of TIMESTAMP in FieldSpec.DataType will be helpful.

Since the stored type is long, we can store any hours/mins/seconds/millis/micros/nanos since epoch  ?

Also, how do we handle timezone here ?

File: pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.queries;
+
+import java.io.File;
+import java.io.IOException;
+import java.sql.Timestamp;
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java:23]: It looks like we should use java.time.*

This article says joda time, java.sql.time* and all of old time/date classes usage should be migrated to use java.time

https://softwareengineering.stackexchange.com/questions/190891/joda-time-vs-java-time

File: pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+  @BeforeClass
+  public void setUp()
+      throws Exception {
+    FileUtils.deleteDirectory(INDEX_DIR);
+
+    List<GenericRow> records = new ArrayList<>(NUM_RECORDS);
+    for (int i = 0; i < NUM_RECORDS; i++) {
+      GenericRow record = new GenericRow();
+      long timestamp = BASE_TIMESTAMP + i;
+      // Insert data in 3 different formats
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java:96]: Can we try couple of values with timezone as well ?

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java [hunk 21]
```diff
@@ -228,6 +229,22 @@ public static long fromEpochDaysBucket(long days, long bucket) {
...
+  @ScalarFunction
+  public static Timestamp toTimestamp(long millis) {
+    return new Timestamp(millis);
+  }
+
+  /**
+   * Converts Timestamp to epoch millis
+   */
+  @ScalarFunction
+  public static long fromTimestamp(Timestamp timestamp) {
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java:244]: (nit) consider naming it fromTimestampToMillisSinceEpoch or something like that ?

Comment (siddharthteotia) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java:244]: Also, what about adding other utility functions like adding/subtracting two timestamps. I think users can probably use ADD/SUB non-scalar transform function on TIMESTAMP type but probably this is a better place

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java [hunk 22]
```diff
@@ -228,6 +229,22 @@ public static long fromEpochDaysBucket(long days, long bucket) {
     return TimeUnit.DAYS.toMillis(days * bucket);
   }
 
+  /**
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java:232]: There are existing functions in this class like dayOfWeek, dayOfYear etc. I think we should add a version of those on Timestamp type. Alternately, since the stored type is long, may be they will automatically work on column of TIMESTAMP type ?

File: pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java (hunk 19 above)
Comment (siddharthteotia) [reply pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java:23]: `Both Joda-Time and the original Java date-time classes are now outmoded by the java.time classes built into Java 8, Java 9, and later.` from the thread

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java (hunk 15 above)
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java:392]: I don't see it very necessary to support bit/byte/short natively. With dictionary encoding, it should perform almost identical to `INT`. If we do want to add them in the future, we can add the metadata then, and make it default to `INT` for backward-compatibility.

File: pinot-core/src/test/java/org/apache/pinot/queries/BooleanQueriesTest.java (hunk 16 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/queries/BooleanQueriesTest.java:102]: Added the value conversion rules into the `PinotDataType` class javadoc

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java (hunk 17 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java:253]: We can add more logical type check in the future, but for existing checks I keep it physical type check so that it is backward-compatible

**Review by siddharthteotia — APPROVED** [review]
LGTM based on our offline discussion

File: pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java (hunk 18 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java:59]: Added more notes in `PinotDataType` which handles the type conversion.
We always store TIMESTAMP in millis since epoch. I don't see much value keeping micros/nanos granularity, and millis is the smallest granularity for time management in pinot

File: pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java (hunk 19 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java:23]: Here we should stick with SQL timestamp format. The `java.sql.Timestamp` is used for ser/de only. Internally we can convert it to `java.time` classes for calculation.

File: pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java (hunk 20 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/queries/TimestampQueriesTest.java:96]: Timezone is not stored in TIMESTAMP type

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java (hunk 21 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java:244]: This follows the naming convention of other functions (e.g. `fromDateTime`, `fromEpochDays` etc.)
We can add more util functions in the following PRs as needed. That should be very straight forward

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java (hunk 22 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java:232]: Yes, it automatically works, and that's one reason why TIMESTAMP is stored as millis since epoch

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java]: Revert this to unsupported because I feel it makes no sense to convert a timestamp to boolean, most likely a human error of putting the wrong data type

**Comment by Jackie-Jiang** [comment]
@MrNeocore TIMESTAMP is internally stored as LONG, so it will be more efficiently performance wise. We will also build optimizations/indexes on TIMESTAMP type, which won't be applied to STRING type because it is too general

**Comment by MrNeocore** [comment]
I see, thanks @Jackie-Jiang
